pkg/istio/client.go

```diff
@@ -69,46 +62,38 @@ func NewVersionedIstioClient(cfg *rest.Config, config Config, recorder record.Ev
 	}, nil
 }
 
-type customClientErr struct {
-	Err        error
-	IsNotFound bool
-}
-
-func (c *Client) getVirtualService(ctx context.Context) (*istioclientapi.VirtualService, *customClientErr) {
+func (c *Client) GetVirtualService(ctx context.Context) (*istioclientapi.VirtualService, error) {
 	virtualService, err := c.NetworkingV1beta1().
 		VirtualServices(metav1.NamespaceDefault).
 		Get(ctx, c.config.VirtualServiceName, metav1.GetOptions{})
-	if client.IgnoreNotFound(err) != nil {
-		return nil, &customClientErr{
-			Err:        fmt.Errorf("failed to fetch virtual service %w", err),
-			IsNotFound: false,
-		}
-	}
-	if apierrors.IsNotFound(err) {
-		return nil, &customClientErr{
-			Err:        err,
-			IsNotFound: true,
-		}
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch virtual service %w", err)
 	}
 	return virtualService, nil
 }
 
-func (c *Client) createVirtualService(ctx context.Context, watcher *v1alpha1.Watcher,
+func (c *Client) CreateVirtualService(ctx context.Context, watcher *v1alpha1.Watcher,
 ) (*istioclientapi.VirtualService, error) {
 	if watcher == nil {
 		return &istioclientapi.VirtualService{}, nil
 	}
 
-	gateway, err := c.lookupGateway(ctx, watcher)
+	virtualSvc := &istioclientapi.VirtualService{}
+	virtualSvc.SetName(c.config.VirtualServiceName)
+	virtualSvc.SetNamespace(metav1.NamespaceDefault)
+
+	gateways, err := c.LookupGateways(ctx, watcher)
 	if err != nil {
 		return nil, err
 	}
+	for i := range gateways {
```

Comment:
Just a style preference, but don't you think it's better to collect all the gateways in the `[]string` slice first, and then just assign the ready slice the `virtualSvc.Spec.Gateways` - once?

Reply:
ok, I also extracted it to a function, please check.